### PR TITLE
Adding OSD cache sizing docs

### LINKED ISSUE #286: Customizing disk_led.sls in DeepSea
**Link To Section**
Please provide a link to the section or chapter you are reporting this bug for.

http://docserv.suse.de/documents/SES_6/ses-admin/html/storage-tips.html#id-1.3.8.2.14.7

**Summary**

The documentation is suggesting to override a file that is shipped by the installation package, because of that all changes will get lost after an update.

**Steps To Reproduce**
Steps to reproduce the behavior.

**Expected Results**

The documentation should explain that it is possible to customize the default behavior of the disk_led.sls by using the DeepSea way (https://github.com/SUSE/DeepSea/blob/master/srv/pillar/ceph/README).

This is done for example here: http://docserv.suse.de/documents/SES_6/ses-admin/html/storage-troubleshooting.html#storage-bp-recover-clockskew

**Actual Results**
Explain what actually happened when the steps were executed, and if applicable, add screenshots to help explain your problem.

**Notes**

The documentation should mention this:

The default behavior can be customized by adding e.g. for the ledmon package and ledctl tool.
``` 
disk_led:
  cmd:
    ident:
      'on': ledctl locate='{device_file}'
      'off': ledctl locate_off='{device_file}'
    fault:
      'on': ledctl locate='{device_file}'
      'off': ledctl locate_off='{device_file}'
```
to `/srv/pillar/ceph/stack/global.yml` or any other YAML file mentioned in https://github.com/SUSE/DeepSea/blob/master/srv/pillar/ceph/README. If the customization should only apply to a special node (minion), then the file `stack/ceph/minions/{{minion}}.yml` needs to be used. Simply reference to the section where this DeepSea feature is explained.

### LINKED ISSUE #299: [doc] 19.2.1  Automatic Cache Sizing
19.2.1  Automatic Cache Sizing

https://susedoc.github.io/doc-ses/master/single-html/ses-admin/#config-auto-cache-sizing

Document does not provide guidance for selecting an osd_memory_target value. This is important, as this selection will further drive hardware requirements.

### PR DESCRIPTION
**Pull request by asettle**
Closes #299

### CODE REVIEW

**Comment by smithfarm** [comment]
Looks good to me. @Martin-Weiss What do you think? Does this change improve the current state, or make it worse?

File: xml/tuning-ceph-tuning.xml [hunk 1]
```diff
@@ -265,17 +265,34 @@ min_alloc_size_hdd=65536
...
-       available for each OSD. The idea being that more low-probability
-       cache hits may occur.
+       Increasing the &bluestore; cache size can improve performance with
+       many workloads. While &filestore; OSD's cache data in the kernel's page
+       cache, &bluestore; OSDs cache data within the memory allocated by the
+       OSD daemon itself. The OSD daemon will allocate memory up to its
+       memory target (as controlled by the <literal>osd_memory_target</literal> parameter),
+       and this determines the potential size of the &bluestore; cache.
+       The &bluestore; cache is a read cache that, by default, is populated
+       by object read. By setting the cache minimum higher than the default,
```
Comment (lproven) [inline xml/tuning-ceph-tuning.xml:278]: "object read" sounds weird to me. 

"when objects are read"?

I also suggest "setting the cache's minimum size".

File: xml/tuning-ceph-tuning.xml [hunk 2]
```diff
@@ -265,17 +265,34 @@ min_alloc_size_hdd=65536
...
+       it is guaranteed that the value specified will be the minimum cache
+       available for each OSD. The idea being that more low-probability cache
+       hits may occur.
+     </para>
+     <para>
+       The default <literal>osd_memory_target</literal> value is 4GB: For example, each OSD daemon
```
Comment (lproven) [inline xml/tuning-ceph-tuning.xml:284]:
```suggestion
       The default <literal>osd_memory_target</literal> value is 4&nbsp;GB: For example, each OSD daemon
```

File: xml/tuning-ceph-tuning.xml [hunk 3]
```diff
@@ -265,17 +265,34 @@ min_alloc_size_hdd=65536
...
+       running on a node can be expected to consume that much memory. If a
+       node's total RAM is significantly higher than <literal>number of OSDs * 4GB</literal>
```
Comment (lproven) [inline xml/tuning-ceph-tuning.xml:286]:
```suggestion
       node's total RAM is significantly higher than <literal>number of OSDs &times; 4&nbsp;GB</literal>
```

File: xml/tuning-ceph-tuning.xml (hunk 1 above)
Comment (asettle) [reply xml/tuning-ceph-tuning.xml:278]: `... object's read`?

Comment (smithfarm) [reply xml/tuning-ceph-tuning.xml:278]: I vote for: "It is populated when objects are read" as the non-engineer equivalent of "It is populated by object read" (which is engineer-speak).

File: xml/tuning-ceph-tuning.xml [hunk 4]
```diff
@@ -265,17 +265,34 @@ min_alloc_size_hdd=65536
...
-       With many workloads, increasing the OSD cache can improve
-       performance. This cache is a read cache that is populated
-       on object read by default. The memory target is inclusive of
-       the base RAM required for the OSD process and the total cache.
-       By setting the cache minimum higher than the default, it is
-       guaranteed that the value specified will be the minimum cache
-       available for each OSD. The idea being that more low-probability
-       cache hits may occur.
+       Increasing the &bluestore; cache size can improve performance with
+       many workloads. While &filestore; OSD's cache data in the kernel's page
```
Comment (lproven) [inline xml/tuning-ceph-tuning.xml:272]:
```suggestion
       many workloads. While &filestore; OSDs cache data in the kernel's page
```

Comment (smithfarm) [reply xml/tuning-ceph-tuning.xml:272]: :+1:

**Comment by Martin-Weiss** [comment]
> 
> 
> Looks good to me. @Martin-Weiss What do you think? Does this change improve the current state, or make it worse?

Is there a rendered version of this documentation changes available where one can read through?
Or how do you prove-read this - in the XML files?

**Comment by asettle** [comment]
> > Looks good to me. @Martin-Weiss What do you think? Does this change improve the current state, or make it worse?
> 
> Is there a rendered version of this documentation changes available where one can read through?
> Or how do you prove-read this - in the XML files?

@Martin-Weiss The XML text is readable, or you can pull the request down yourself and build using DAPs. If you are unable to do either, let me know and I will build a PDF for you. But as this is a text inclusion, the xml at this point should be fine :) : https://github.com/SUSE/doc-ses/pull/323/files#diff-6bbfb44b26ddfc527ec621ffca0e9f83R271

**Comment by Martin-Weiss** [comment]
> 
> 
> > > Looks good to me. @Martin-Weiss What do you think? Does this change improve the current state, or make it worse?
> > 
> > 
> > Is there a rendered version of this documentation changes available where one can read through?
> > Or how do you prove-read this - in the XML files?
> 
> @Martin-Weiss The XML text is readable, or you can pull the request down yourself and build using DAPs. If you are unable to do either, let me know and I will build a PDF for you. But as this is a text inclusion, the xml at this point should be fine :) : https://github.com/SUSE/doc-ses/pull/323/files#diff-6bbfb44b26ddfc527ec621ffca0e9f83R271

It would be nice if we could publish latest changes for SES similar to https://susedoc.github.io/doc-caasp/master/single-html/caasp-admin/ ;-).  
The easier we publish - the easier to get updates and feedback ;-).

Just had a look at https://github.com/SUSE/doc-ses/pull/323/files#diff-6bbfb44b26ddfc527ec621ffca0e9f83R271 - looks good from my point of view. I just would add the best practice recommendation - take memory - remove ~16  GB or 32 GB depending on whats running on the OS side, remove MDS cache (in case MDS is also running on that machine) and then divide the rest by the number of OSDs on that host and leave some room for improvement..

So 256 GB in the server for 20 OSDs, no other daemons -> (256 GB - 32 GB ) / 20 OSDs = 11,2 GB/OSD (max) -> here I would configure something around 8 to 10 GB per OSD.

**Comment by asettle** [comment]
> It would be nice if we could publish latest changes for SES similar to https://susedoc.github.io/doc-caasp/master/single-html/caasp-admin/ ;-).
> The easier we publish - the easier to get updates and feedback ;-).
> 
This is available for every product we document, including SES. You can see it here: https://susedoc.github.io/#ses

However, this is only updated when the pull request has been approved and merged. This is an unmerged pull request you are viewing. 

> Just had a look at https://github.com/SUSE/doc-ses/pull/323/files#diff-6bbfb44b26ddfc527ec621ffca0e9f83R271 - looks good from my point of view. I just would add the best practice recommendation - take memory - remove ~16 GB or 32 GB depending on whats running on the OS side, remove MDS cache (in case MDS is also running on that machine) and then divide the rest by the number of OSDs on that host and leave some room for improvement..
> 
> So 256 GB in the server for 20 OSDs, no other daemons -> (256 GB - 32 GB ) / 20 OSDs = 11,2 GB/OSD (max) -> here I would configure something around 8 to 10 GB per OSD.

Thanks.

**Comment by Martin-Weiss** [comment]
> > It would be nice if we could publish latest changes for SES similar to https://susedoc.github.io/doc-caasp/master/single-html/caasp-admin/ ;-).
> > The easier we publish - the easier to get updates and feedback ;-).
> 
> This is available for every product we document, including SES. You can see it here: https://susedoc.github.io/#ses
> 
> However, this is only updated when the pull request has been approved and merged. This is an unmerged pull request you are viewing.
> 

Hm - do you think we could also create one additional version "https://susedoc.github.io/#ses-pre-official-merge" that has all the latest pull requests merged?

**Comment by asettle** [comment]
> > > It would be nice if we could publish latest changes for SES similar to https://susedoc.github.io/doc-caasp/master/single-html/caasp-admin/ ;-).
> > > The easier we publish - the easier to get updates and feedback ;-).
> > 
> > 
> > This is available for every product we document, including SES. You can see it here: https://susedoc.github.io/#ses
> > However, this is only updated when the pull request has been approved and merged. This is an unmerged pull request you are viewing.
> 
> Hm - do you think we could also create one additional version "https://susedoc.github.io/#ses-pre-official-merge" that has all the latest pull requests merged?

All of the latest _merged_ pull requests are available in the relevant branches. For example, this change will be available in the _master_ branch, once it is approved and I merged it. This is an _unmerged_ pull request currently. If you are looking for a preview of the content before the branch is merged, this is a larger request that I cannot officially handle. It has been brought up before with @sknorr

**Comment by Martin-Weiss** [comment]
@sknorr  - do you think we can create an other branch that auto-merges pull requests and provides a pre-official-merge rendered view?

**Comment by asettle** [comment]
> @sknorr - do you think we can create an other branch that auto-merges pull requests and provides a pre-official-merge rendered view?

Sorry Martin, this is not something we want. Pull requests are a part of our development process and are a requirement for proper review standards in our team - this pull request is created purposefully. If you would like a *preview* on the pull request, that is something to discuss with @sknorr :)

**Comment by sknorr** [comment]
Hey -- so, the quick solution is that either of you (Martin or Alex) just add the feature branch for this PR [here](https://github.com/SUSEdoc/susedoc.github.io/blob/master/config.xml#L144,L148)  for the documents needed. This gives you a preview document, just like for `master` etc. If you like, Martin, I can give you push access to the repo too.

In general: I looked at what currently stops us from just building all branches of doc repos, and it's not too much. But it definitely is not something I can afford to tackle this week either. If you want to help out with the implementation, I'll be happy to give you pointers.

**Comment by asettle** [comment]
@smithfarm @ifed01 - if you both have some time today, a last review would be much appreciated :) I've added Martin's example in.

File: xml/tuning-ceph-tuning.xml [hunk 5]
```diff
@@ -298,12 +298,25 @@ min_alloc_size_hdd=65536
 osd_memory_target = 6442450944
 osd_memory_cache_min = 4294967296
 </screen>
+     <tip>
+       <para>
+         As a best practice, adjust the memory by removing 16&nbsp;GB or 32&nbsp;GB
+         (depending on the OS), remove the MDS cache, and divide the remainder
+         by the number of OSDs on that host. Ensure you leave room for improvement.
```
Comment (smithfarm) [inline xml/tuning-ceph-tuning.xml:305]: As a best practice, start with the full memory of the node. First, deduct 16 GB - 32 GB for the OS. Second, deduct appropriate amounts for any other workloads running on the node (e.g. MDS cache if MDS is colocated). Divide the remainder . . .

**Pull request merged**